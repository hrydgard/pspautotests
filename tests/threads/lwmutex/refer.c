#include "shared.h"

void testRefer(const char *title, SceLwMutexWorkarea *workarea, SceKernelLwMutexInfo *info) {
	int result = sceKernelReferLwMutexStatus(workarea, info);
	if (result == 0) {
		schedf("%s: ", title);
		schedfLwMutexInfo(workarea, info);
	} else {
		schedf("%s: Failed (%X)\n", title, result);
	}
}

void testReferByID(const char *title, SceUID uid, SceKernelLwMutexInfo *info) {
	int result = sceKernelReferLwMutexStatusByID(uid, info);
	if (result == 0) {
		schedf("%s: ", title);
		schedfLwMutexInfo((SceLwMutexWorkarea *)info->workarea, info);
	} else {
		schedf("%s: Failed (%X)\n", title, result);
	}
}

int main(int argc, char **argv) {
	SceLwMutexWorkarea workarea;
	SceKernelLwMutexInfo info;

	schedf("sceKernelReferLwMutexStatus:\n");
	sceKernelCreateLwMutex(&workarea, "refer", 0, 0, NULL);
	info.size = sizeof(info);

	// Crashes.
	//testRefer("NULL info", &workarea, NULL);
	testRefer("Normal", &workarea, &info);

	schedf("\nSizes:\n");
	int sizes[] = {0, 1, 2, 4, 8, 44, 48, 52, 56, 64, 128, 256, 1024, 0xFF00, 0xFF0000, 0x80000001, 0xCCCCCCCC, 0xFFFFFFFF};
	int i, result;
	for (i = 0; i < sizeof(sizes) / sizeof(sizes[0]); ++i) {
		info.size = sizes[i];
		result = sceKernelReferLwMutexStatus(&workarea, &info);
		schedf("  %08X => %08X (result=%08X)\n", sizes[i], info.size, result);
	}
	schedf("\n");
	flushschedf();

	sceKernelDeleteLwMutex(&workarea);

	// Crashes.
	//testRefer("NULL", 0, &info);
	//testRefer("Invalid", (SceLwMutexWorkarea *)0xDEADBEEF, &info);
	testRefer("Deleted", &workarea, &info);

	BASIC_SCHED_TEST("Refer other",
		result = sceKernelReferLwMutexStatus(&workarea2, &info);
	);
	BASIC_SCHED_TEST("Refer same",
		result = sceKernelReferLwMutexStatus(&workarea1, &info);
	);

	schedf("\n");
	schedf("sceKernelReferLwMutexStatusByID:\n");
	sceKernelCreateLwMutex(&workarea, "refer", 0, 0, NULL);
	info.size = sizeof(info);

	// Crashes.
	//testReferByID("NULL info", workarea.uid, NULL);
	testReferByID("Normal", workarea.uid, &info);

	schedf("\nSizes:\n");
	for (i = 0; i < sizeof(sizes) / sizeof(sizes[0]); ++i) {
		info.size = sizes[i];
		result = sceKernelReferLwMutexStatusByID(workarea.uid, &info);
		schedf("  %08X => %08X (result=%08X)\n", sizes[i], info.size, result);
	}
	schedf("\n");
	flushschedf();

	sceKernelDeleteLwMutex(&workarea);

	testReferByID("NULL", 0, &info);
	testReferByID("Invalid", 0xDEADBEEF, &info);
	testReferByID("Deleted", workarea.uid, &info);

	BASIC_SCHED_TEST("Refer other",
		result = sceKernelReferLwMutexStatusByID(workarea2.uid, &info);
	);
	BASIC_SCHED_TEST("Refer same",
		result = sceKernelReferLwMutexStatusByID(workarea1.uid, &info);
	);

	flushschedf();
	return 0;
}