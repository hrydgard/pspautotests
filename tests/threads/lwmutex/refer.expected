sceKernelReferLwMutexStatus:
Normal: LwMutex: OK (size=64,name=refer,attr=00000000,uid=1,workarea=1,init=0,current=0,lockThread=0)
LwMutexWorkarea (count=0, thread=00000000, attr=000, waiting=0, uid=1, 00000000, 00000000, 00000000)

Sizes:
  00000000 => 00000000 (result=00000000)
  00000001 => 00000040 (result=00000000)
  00000002 => 00000040 (result=00000000)
  00000004 => 00000040 (result=00000000)
  00000008 => 00000040 (result=00000000)
  0000002C => 00000040 (result=00000000)
  00000030 => 00000040 (result=00000000)
  00000034 => 00000040 (result=00000000)
  00000038 => 00000040 (result=00000000)
  00000040 => 00000040 (result=00000000)
  00000080 => 00000040 (result=00000000)
  00000100 => 00000040 (result=00000000)
  00000400 => 00000040 (result=00000000)
  0000FF00 => 00000040 (result=00000000)
  00FF0000 => 00000040 (result=00000000)
  80000001 => 00000040 (result=00000000)
  CCCCCCCC => 00000040 (result=00000000)
  FFFFFFFF => 00000040 (result=00000000)

Deleted: Failed (800201CA)
Refer other: A1B1C2E1D1F2 (thread=800201B5, main=00000000)
Refer same: A1B1C2E1D1F2 (thread=800201B5, main=00000000)

sceKernelReferLwMutexStatusByID:
Normal: LwMutex: OK (size=64,name=refer,attr=00000000,uid=1,workarea=1,init=0,current=0,lockThread=0)
LwMutexWorkarea (count=0, thread=00000000, attr=000, waiting=0, uid=1, 00000000, 00000000, 00000000)

Sizes:
  00000000 => 00000000 (result=00000000)
  00000001 => 00000040 (result=00000000)
  00000002 => 00000040 (result=00000000)
  00000004 => 00000040 (result=00000000)
  00000008 => 00000040 (result=00000000)
  0000002C => 00000040 (result=00000000)
  00000030 => 00000040 (result=00000000)
  00000034 => 00000040 (result=00000000)
  00000038 => 00000040 (result=00000000)
  00000040 => 00000040 (result=00000000)
  00000080 => 00000040 (result=00000000)
  00000100 => 00000040 (result=00000000)
  00000400 => 00000040 (result=00000000)
  0000FF00 => 00000040 (result=00000000)
  00FF0000 => 00000040 (result=00000000)
  80000001 => 00000040 (result=00000000)
  CCCCCCCC => 00000040 (result=00000000)
  FFFFFFFF => 00000040 (result=00000000)

NULL: Failed (800201CA)
Invalid: Failed (800201CA)
Deleted: Failed (800201CA)
Refer other: A1B1C2E1D1F2 (thread=800201B5, main=00000000)
Refer same: A1B1C2E1D1F2 (thread=800201B5, main=00000000)
